src/Nest/Indices/MappingManagement/GetMapping/GetMappingResponse.cs

```diff
@@ -20,39 +18,30 @@ internal class GetRootObjectMappingWrapping : Dictionary<string, Dictionary<stri
 
 	public class GetMappingResponse : ResponseBase, IGetMappingResponse
 	{
-		public override bool IsValid => base.IsValid && this.Mapping != null;
-
 		internal GetMappingResponse() { }
 
-		internal GetMappingResponse(IApiCallDetails status, GetRootObjectMappingWrapping dict)
+		internal GetMappingResponse(GetRootObjectMappingWrapping dict)
 		{
-			this.Mappings = new Dictionary<string, IList<TypeMapping>>();
-			//TODO can dict truely ever be null, whats the response look like when field mapping is not found.
-			//does status.Success not already reflect this?
-			//this.IsValid = status.Success && dict != null && dict.Count > 0;
-			if (!status.Success || dict == null) return;
-
 			foreach (var index in dict)
 			{
-				if (index.Value == null || !index.Value.ContainsKey("mappings"))
-					continue;
-
-				var mappings = index.Value["mappings"];
-				this.Mappings.Add(index.Key, new List<TypeMapping>());
-				if (mappings == null) continue;
-				foreach (var mapping in mappings)
+				Dictionary<string, TypeMapping> mappings;
+				if (index.Value != null && index.Value.TryGetValue("mappings", out mappings))
 				{
-					if (mapping.Value == null) continue;
-					this.Mappings[index.Key].Add(mapping.Value);
+					this.Mappings.Add(index.Key, new Dictionary<TypeName, TypeMapping>());
+					foreach (var mapping in mappings)
+					{
+						if (mapping.Value == null) continue;
+						this.Mappings[index.Key].Add(mapping.Key, mapping.Value);
+					}
 				}
 			}
-			
-			this.Mapping = this.Mappings.Where(kv=>kv.Value.HasAny(v=>v != null))
-				.SelectMany(kv=>kv.Value)
-				.FirstOrDefault(t=>t != null);
+
+			this.Mapping = this.Mappings.Where(kv => kv.Value.HasAny(v => v.Value != null))
+				.SelectMany(kv => kv.Value)
+				.FirstOrDefault(t => t.Value != null).Value;
 		}
 
-		public Dictionary<string, IList<TypeMapping>> Mappings { get; internal set; } = new Dictionary<string, IList<TypeMapping>>();
+		public Dictionary<IndexName, IDictionary<TypeName, TypeMapping>> Mappings { get; internal set; } = new Dictionary<IndexName, IDictionary<TypeName, TypeMapping>>();
```

Comment:
if we support `IndexName` and `TypeName` on responses we also have to support the following use case:

```csharp
var index = Index<Project>();
var mapping = this.client.GetMapping(new GetMappingRequest(index, AllTypes));
var indexMappings = mapping.Mappings[index];
```

Since the response `Index` is build from `string` and the request `index` is an unresolved `IndexName` it won't be found where you would totally expect it to.

I would love to support this but in 2.x we gave up because we would have to inject ConnectionSettings as state of all responses. 

Ideas here welcome! :smile: